integrations/unftp-sbe/src/lib.rs

```diff
@@ -14,3 +14,221 @@
+        })
+    }
+
+    fn gid(&self) -> u32 {
+        0
+    }
+
+    fn uid(&self) -> u32 {
+        0
+    }
+}
+
+fn cvt_err(err: opendal::Error) -> storage::Error {
+    let kind = match err.kind() {
+        opendal::ErrorKind::NotFound => storage::ErrorKind::PermanentFileNotAvailable,
+        opendal::ErrorKind::AlreadyExists => storage::ErrorKind::PermanentFileNotAvailable,
+        opendal::ErrorKind::PermissionDenied => storage::ErrorKind::PermissionDenied,
+        _ => storage::ErrorKind::LocalError,
+    };
+    storage::Error::new(kind, err)
+}
+
+fn cvt_path(path: &Path) -> storage::Result<&str> {
+    path.to_str().ok_or_else(|| {
+        storage::Error::new(
+            storage::ErrorKind::LocalError,
+            "Path is not a valid UTF-8 string",
+        )
+    })
+}
+
+#[async_trait::async_trait]
+impl<User: UserDetail> StorageBackend<User> for OpendalStorage {
+    type Metadata = OpendalMetadata;
+
+    async fn metadata<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+    ) -> storage::Result<Self::Metadata> {
+        let metadata = self
+            .op
+            .stat(cvt_path(path.as_ref())?)
+            .await
+            .map_err(cvt_err)?;
+        Ok(OpendalMetadata(metadata))
+    }
+
+    async fn list<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+    ) -> storage::Result<Vec<storage::Fileinfo<PathBuf, Self::Metadata>>>
+    where
+        Self::Metadata: storage::Metadata,
+    {
+        let ret = self
+            .op
+            .list(cvt_path(path.as_ref())?)
+            .await
+            .map_err(cvt_err)?
+            .into_iter()
+            .map(|x| {
+                let (path, metadata) = x.into_parts();
+                storage::Fileinfo {
+                    path: path.into(),
+                    metadata: OpendalMetadata(metadata),
+                }
+            })
+            .collect();
+        Ok(ret)
+    }
+
+    async fn get<P: AsRef<Path> + Send + Debug>(
+        &self,
+        _: &User,
+        path: P,
+        start_pos: u64,
+    ) -> storage::Result<Box<dyn tokio::io::AsyncRead + Send + Sync + Unpin>> {
+        let buf = self
+            .op
+            .read_with(cvt_path(path.as_ref())?)
+            .range(start_pos..)
+            .await
+            .map_err(cvt_err)?;
+        Ok(Box::new(IoBuffer(buf)))
+    }
+
+    async fn put<
+        P: AsRef<Path> + Send + Debug,
+        R: tokio::io::AsyncRead + Send + Sync + Unpin + 'static,
+    >(
+        &self,
+        _: &User,
+        mut input: R,
+        path: P,
+        _: u64,
+    ) -> storage::Result<u64> {
+        let mut buf = self.pool.get();
+        input.read_buf(&mut buf).await?;
+        let bytes = buf.split().freeze();
+        let len = bytes.len() as u64;
+        self.pool.put(buf);
+        self.op
+            .write_with(cvt_path(path.as_ref())?, bytes)
```

Comment:
Please use `op.writer()` and `into_futures_async_write` instead of reading all data in memory first.